parasol/solr/client.py

```diff
@@ -0,0 +1,131 @@
+import json
+import logging
+import sys
+import time
+from urllib.parse import urljoin
+
+from attrdict import AttrDict
+import requests
+
+# Start making Django agnostic from the get-go
+try:
+    from django.settings import settings
+    from django.core.exceptions import ImproperlyConfigured
+except ImportError:
+    pass
+
+
+from parasol import __version__ as parasol_version
+from parasol.solr.schema import Schema
+from parasol.solr.update import Update
+from parasol.solr.admin import CoreAdmin
+
+logger = logging.getLogger(__name__)
+
+
+
+class SolrClientException(Exception):
+    '''Base class for all exceptions in this module'''
+    pass
+
+
+class CoreForTestExists(SolrClientException):
+    '''Raised when default core for running unit tests exists'''
+
+
+
+class SolrClient:
+    '''Base class for all SolrClient with sane development defaults'''
+    #: Url for solr base instance
+    solr_url = 'http://localhost:8983/solr'
+    #: core admin handler
+    core_admin_handler = 'admin/cores'
+    #: select handler
+    select_handler = 'select'
+    #: schema handler
+    schema_handler = 'schema'
+    # update handler
+    update_handler = 'update'
+    #: core or collection
+    collection = ''
+    # commitWithin definition
+    commit_within = 1000
+
+    def __init__(self, *args, **kwargs):
+        # For now, a generous init that will override valuesbut not
+        # choke on an unexpected one.
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+        # Don't allow session configuration to be so easily overwritten
+        self.session = requests.Session()
+        self.session.headers = {
+            'User-Agent': 'Parasol/%s' % parasol_version
+        }
+
+        self.schema = Schema(self)
+        self.update = Update(self, commitWithin=1000)
+        self.core_admin = CoreAdmin(self)
+
+    def build_url(self, handler):
+        '''Return a url to a handler based on core and base url'''
+        # Two step proecess to avoid quirks in urljoin behavior
+        # First, join the collection/core with a slashes appended
+        # just in case so # it doesn't ovewrite the base url
+        # (extras cause no issues)
+        collection = urljoin('%s/' % self.solr_url, '%s/' % self.collection)
+        # then return the core joined with handler -- without slash per
+        # Solr API docs.
+        return urljoin(collection, handler)
+
+    def make_request(self, meth, url, headers={},
+                      params={}, data=None, **kwargs):
+        '''Private method for making a request to Solr, wraps session.request'''
+        # always add wt=json for JSON api
+        if 'wt' not in params:
+            params.update({'wt': 'json'})
+        start = time.time()
+        response = self.session.request(
+            meth,
+            url,
+            params=params,
+            headers=headers,
+            data=json.dumps(data),
+            **kwargs)
+        # log the time as needed
```

Comment:
Debug logging should always be output here regardless of response code. That should allow you to simplify the code - no need to store total time or log string.

Reply:
Agreed, and cleaned up now.